### 조선 일보 데이터크롤러

```
Arguments : query_word, (period)

return : title_list, category_list, body_list
```

In [1]:
import requests
from bs4 import BeautifulSoup

In [2]:
def body_extractor(link):
    soup = BeautifulSoup(requests.get(link).content, 'html.parser')
    
    text = ""
    for t in soup.select(".par"):
        text += " " + t.text

    return text.strip()

def chosun_crawler(query, period=None):
    
    # data containers 
    titles = []
    links = []
    categories = []
    dates = []
    bodies = []
    
    i = 1 
    nobody = 0
    
    while True:
        print("-" * 30)
        print('{} page is start'.format(i))
        
        url = "http://search.chosun.com/search/news.search?query={}&pageno={}&orderby=news".format(query, i)
        res = requests.get(url)
        soup = BeautifulSoup(res.content, 'html.parser')
        
        if soup.select(".result_none"):
            print('***** no result! *****')
            break

        for art, info in zip(soup.select("dt > a"), soup.select(".art_info")):
            
            title = art.text
            if len(title) == 0:
                nobody += 1
                continue
                
            link = art["href"]
            if len(link) == 0:
                nobody += 1
                continue
                
            category = info.select_one(".bread_crumbs").text
            if len(category) == 0:
                category = 'unknown_category'
                
            date = info.select_one(".date").text
            if len(date) == 0:
                date = "unknown_date"
            
            try:
                body = body_extractor(link)
                if len(body) == 0:
                    print("no body")
                    nobody += 1
                    continue
                    
            except Exception as ex:
                print(ex)
                nobody += 1
                continue
                
            titles.append(title)
            links.append(link)
            categories.append(category) 
            dates.append(date)
            bodies.append(body)
            
        print('{} page is done'.format(i))
        i += 1
    return [titles, links, categories, dates, bodies], nobody

In [ ]:
data_dust, nobody_dust = chosun_crawler('미세먼지')

------------------------------
1 page is start
1 page is done
------------------------------
2 page is start
2 page is done
------------------------------
3 page is start
3 page is done
------------------------------
4 page is start
4 page is done
------------------------------
5 page is start
5 page is done
------------------------------
6 page is start
6 page is done
------------------------------
7 page is start
7 page is done
------------------------------
8 page is start
8 page is done
------------------------------
9 page is start
9 page is done
------------------------------
10 page is start
10 page is done
------------------------------
11 page is start
11 page is done
------------------------------
12 page is start
12 page is done
------------------------------
13 page is start
13 page is done
------------------------------
14 page is start
no body
no body
14 page is done
------------------------------
15 page is start
15 page is done
------------------------------
16 page is 

126 page is done
------------------------------
127 page is start
no body
127 page is done
------------------------------
128 page is start
128 page is done
------------------------------
129 page is start
no body
no body
no body
129 page is done
------------------------------
130 page is start
130 page is done
------------------------------
131 page is start
no body
no body
131 page is done
------------------------------
132 page is start
no body
132 page is done
------------------------------
133 page is start
133 page is done
------------------------------
134 page is start
no body
134 page is done
------------------------------
135 page is start
135 page is done
------------------------------
136 page is start
no body
136 page is done
------------------------------
137 page is start
137 page is done
------------------------------
138 page is start
138 page is done
------------------------------
139 page is start
139 page is done
------------------------------
140 page is start
140 

246 page is done
------------------------------
247 page is start
247 page is done
------------------------------
248 page is start
248 page is done
------------------------------
249 page is start
249 page is done
------------------------------
250 page is start
250 page is done
------------------------------
251 page is start
251 page is done
------------------------------
252 page is start
252 page is done
------------------------------
253 page is start
253 page is done
------------------------------
254 page is start
254 page is done
------------------------------
255 page is start
no body
255 page is done
------------------------------
256 page is start
256 page is done
------------------------------
257 page is start
no body
257 page is done
------------------------------
258 page is start
258 page is done
------------------------------
259 page is start
259 page is done
------------------------------
260 page is start
260 page is done
------------------------------
261 page is 

367 page is done
------------------------------
368 page is start
368 page is done
------------------------------
369 page is start
no body
369 page is done
------------------------------
370 page is start
370 page is done
------------------------------
371 page is start
371 page is done
------------------------------
372 page is start
372 page is done
------------------------------
373 page is start
373 page is done
------------------------------
374 page is start
374 page is done
------------------------------
375 page is start
no body
375 page is done
------------------------------
376 page is start
no body
376 page is done
------------------------------
377 page is start
no body
377 page is done
------------------------------
378 page is start
378 page is done
------------------------------
379 page is start
379 page is done
------------------------------
380 page is start
380 page is done
------------------------------
381 page is start
no body
no body
381 page is done
-----------

483 page is done
------------------------------
484 page is start
no body
484 page is done
------------------------------
485 page is start
485 page is done
------------------------------
486 page is start
486 page is done
------------------------------
487 page is start
487 page is done
------------------------------
488 page is start
no body
488 page is done
------------------------------
489 page is start
489 page is done
------------------------------
490 page is start
no body
no body
490 page is done
------------------------------
491 page is start
491 page is done
------------------------------
492 page is start
492 page is done
------------------------------
493 page is start
493 page is done
------------------------------
494 page is start
no body
494 page is done
------------------------------
495 page is start
495 page is done
------------------------------
496 page is start
496 page is done
------------------------------
497 page is start
no body
497 page is done
-----------

596 page is done
------------------------------
597 page is start
597 page is done
------------------------------
598 page is start
no body
598 page is done
------------------------------
599 page is start
599 page is done
------------------------------
600 page is start
no body
600 page is done
------------------------------
601 page is start
601 page is done
------------------------------
602 page is start
no body
602 page is done
------------------------------
603 page is start
no body
603 page is done
------------------------------
604 page is start
no body
604 page is done
------------------------------
605 page is start
605 page is done
------------------------------
606 page is start
606 page is done
------------------------------
607 page is start
no body
607 page is done
------------------------------
608 page is start
no body
608 page is done
------------------------------
609 page is start
no body
609 page is done
------------------------------
610 page is start
610 page is 

715 page is done
------------------------------
716 page is start
716 page is done
------------------------------
717 page is start
717 page is done
------------------------------
718 page is start
718 page is done
------------------------------
719 page is start
no body
719 page is done
------------------------------
720 page is start
720 page is done
------------------------------
721 page is start
721 page is done
------------------------------
722 page is start
722 page is done
------------------------------
723 page is start
723 page is done
------------------------------
724 page is start
724 page is done
------------------------------
725 page is start
725 page is done
------------------------------
726 page is start
no body
726 page is done
------------------------------
727 page is start
727 page is done
------------------------------
728 page is start
728 page is done
------------------------------
729 page is start
729 page is done
------------------------------
730 page is 

809 page is done
------------------------------
810 page is start
810 page is done
------------------------------
811 page is start
HTTPConnectionPool(host='car.biz.chosun.com', port=80): Max retries exceeded with url: /site/data/html_dir/2014/03/25/2014032504056.html (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x1c1d37ad30>: Failed to establish a new connection: [Errno 61] Connection refused',))
811 page is done
------------------------------
812 page is start
no body
812 page is done
------------------------------
813 page is start
813 page is done
------------------------------
814 page is start
814 page is done
------------------------------
815 page is start
815 page is done
------------------------------
816 page is start
816 page is done
------------------------------
817 page is start
817 page is done
------------------------------
818 page is start
818 page is done
------------------------------
819 page is start
819 page is done
---------------

##### 1st 시도
-  page : 253
-  nobody_dust : 59

##### 2nd 시도

- page : 861
- Memory error